File: go/parsers/invocation.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Invocation struct {
	gen *gen.Invocation
	ex  *PayloadExpander
}

func NewInvocationFields(g *gen.Invocation, payload []byte) messages.InvocationFields {
	return &Invocation{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.PayloadSerializerID()},
	}
}

func (i *Invocation) RequestID() uint64 {
	return i.gen.RequestID()
}

func (i *Invocation) RegistrationID() uint64 {
	return i.gen.RegistrationID()
}

func (i *Invocation) Args() []any {
	return i.ex.Args()
}

func (i *Invocation) KwArgs() map[string]any {
	return i.ex.Kwargs()
}

func (i *Invocation) Details() map[string]any {
	var details map[string]any

	if i.gen.Caller() > 0 {
		setDetail(&details, "caller", i.gen.Caller())
	}

	if i.gen.HasCallerAuthID() {
		authID, _ := i.gen.CallerAuthID()
		setDetail(&details, "caller_authid", authID)
	}

	if i.gen.HasCallerAuthRole() {
		authRole, _ := i.gen.CallerAuthRole()
		setDetail(&details, "caller_authrole", authRole)
	}

	if i.gen.HasProcedure() {
		topic, _ := i.gen.Procedure()
		setDetail(&details, "procedure", topic)
	}

	return details
}

func (i *Invocation) PayloadIsBinary() bool {
	return true
}

func (i *Invocation) Payload() []byte {
	return i.ex.Payload()
}

func (i *Invocation) PayloadSerializer() uint64 {
	return i.gen.PayloadSerializerID()
}

func InvocationToCapnproto(m *messages.Invocation) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	invocation, err := gen.NewRootInvocation(seg)
	if err != nil {
		return nil, err
	}

	invocation.SetRequestID(m.RequestID())
	invocation.SetRegistrationID(m.RegistrationID())
	invocation.SetPayloadSerializerID(serializers.MsgPackSerializerID)

	payload, err := Encode(serializers.MsgPackSerializerID, m.Args(), m.KwArgs())
	if err != nil {
		return nil, err
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeInvocation, data, payload), nil
}

func CapnprotoToInvocation(data, payload []byte) (*messages.Invocation, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	invocation, err := gen.ReadRootInvocation(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewInvocationWithFields(NewInvocationFields(&invocation, payload)), nil
}

File: python/wamp_msgs_capnp/parsers/invocation.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.invocation import Invocation, IInvocationFields
from wampproto.serializers.payload import serialize_payload

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "invocation.capnp")
invocation_capnp = capnp.load(str(schema_file))


class InvocationFields(IInvocationFields):
    def __init__(self, gen, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.payloadSerializerID)

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def registration_id(self) -> str:
        return self._gen.registrationID

    @property
    def args(self) -> list[Any]:
        return self._ex.args

    @property
    def kwargs(self) -> dict[str, Any]:
        return self._ex.kwargs

    @property
    def details(self) -> dict[str, Any]:
        details = {}

        if self._gen.caller:
            details["caller"] = self._gen.caller

        if self._gen.callerAuthID:
            details["caller_authid"] = self._gen.callerAuthID

        if self._gen.callerAuthRole:
            details["caller_authrole"] = self._gen.callerAuthRole

        if self._gen.procedure:
            details["procedure"] = self._gen.procedure

        return details

    @property
    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes:
        return self._ex.payload

    @property
    def payload_serializer(self) -> int:
        return self._gen.payloadSerializerID


def invocation_to_capnproto(i: Invocation) -> bytes:
    invocation = invocation_capnp.Invocation.new_message()

    invocation.requestID = i.request_id
    invocation.registrationID = i.registration_id

    payload_serializer = helpers.select_payload_serializer(i.details)
    invocation.payloadSerializerID = payload_serializer

    if i.payload_is_binary():
        payload = i.payload
    else:
        payload = serialize_payload(payload_serializer, i.args, i.kwargs)

    packed_data = invocation.to_bytes_packed()

    return helpers.prepend_header(Invocation.TYPE, packed_data, payload)


def capnproto_to_invocation(data: bytes) -> Invocation:
    message_data, payload_data = helpers.extract_message(data)
    invocation_obj = invocation_capnp.Invocation.from_bytes_packed(message_data)

    return Invocation(InvocationFields(invocation_obj, payload_data))
